include/range/v3/view/shared.hpp

```diff
@@ -0,0 +1,148 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Filip Matzner 2017
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+#ifndef RANGES_V3_VIEW_SHARED_HPP
+#define RANGES_V3_VIEW_SHARED_HPP
+
+#include <memory>
+#include <type_traits>
+#include <meta/meta.hpp>
+#include <range/v3/begin_end.hpp>
+#include <range/v3/iterator_range.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/size.hpp>
+#include <range/v3/view/all.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        template<typename Rng>
+        struct shared_view
+          : view_interface<shared_view<Rng>,
+            range_cardinality<Rng>::value>
+        {
+        private:
+            // shared storage
+            std::shared_ptr<Rng> rng_ptr_;
+
+        public:
+            shared_view() = default;
+
+            // construct from a shared_ptr
+            shared_view(std::shared_ptr<Rng> pt) noexcept
+              : rng_ptr_{std::move(pt)}
+            {}
+
+            // construct from a range rvalue
+            explicit shared_view(Rng && t)
+              : rng_ptr_{std::make_shared<Rng>(std::move(t))}
+            {}
+
+            // use the stored range's begin and end
+            iterator_t<Rng> begin() const
+            {
+                return ranges::begin(*rng_ptr_);
+            }
+            sentinel_t<Rng> end() const
+            {
+                return ranges::end(*rng_ptr_);
+            }
+
+            // use the const-most size() function provided by the range
+            CONCEPT_REQUIRES(SizedRange<const Rng>())
+            range_size_type_t<Rng> size() const
+            {
+                return ranges::size(*rng_ptr_);
+            }
+            CONCEPT_REQUIRES(SizedRange<Rng>() && !SizedRange<const Rng>())
+            range_size_type_t<Rng> size()
+            {
+                return ranges::size(*rng_ptr_);
+            }
+
+            // shared storage access
+            const std::shared_ptr<Rng> & get_shared() const
+            {
+                return rng_ptr_;
+            }
+        };
+
+        /// \relates all
+        /// \addtogroup group-views
+        /// @{
+        namespace view
+        {
+            struct shared_fn : pipeable<shared_fn>
+            {
+            public:
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(Range<Rng>())>
+                shared_view<Rng> operator()(std::shared_ptr<Rng> pt) const
+                {
+                    return {std::move(pt)};
+                }
```

Comment:
ok, done. I thought that it could come in handy when I already have a shared_ptr to a range in hand. However, if the range is efficiently move constructible, it should not affect the performance if it is moved instead of passing the shared_ptr.